User config flow and custom panel for Dynalite integration

ziv1234 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
The change deprecates the YAML configuration. it imports it into the UI and adds a repair issue, so it won't break a user configuration or change the existing behavior, but since it adds a "deprecation" warning, better to be explicit

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The Dynalite configuration process is done today via YAML which we want to deprecate. However, the configuration is complex since it involved several items (Dynalite has no ability to query, so config has to be set manually).
This adds a user config flow to setup the basics (host / port). After this, a panel appears that can configure the various config entries in a straightforward interactive way

A few example screenshots:

Viewing all defined DYNET areas:
![image](https://user-images.githubusercontent.com/16467659/186098138-f7368be5-03d0-4f18-9dd9-596b36d385fa.png)

Configuring the system global settings:
![image](https://user-images.githubusercontent.com/16467659/186098374-3b307d2e-9590-4c39-bd3a-ef9a7daa7003.png)

Editing a specific DYNET area:
![image](https://user-images.githubusercontent.com/16467659/186098614-ba0c6c87-64dc-41d2-9aa6-9e95b11f729c.png)

Editing a DYNET channel:
![image](https://user-images.githubusercontent.com/16467659/186098674-2aabe374-af27-4bce-8ad4-350c0511fb2f.png)

Defining a new DYNET preset:
![image](https://user-images.githubusercontent.com/16467659/186098766-47f2ed96-1ec7-4397-9a39-cdf0d037b99f.png)

Switching between different systems (multiple config entries)
![image](https://user-images.githubusercontent.com/16467659/186098877-0992101a-d727-40bc-a4f2-6b2d62b345fb.png)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

ziv1234 committed:
first semi-working version

ziv1234 committed:
working and getting the config

ziv1234 committed:
first working version
refactored the schema to its own file

ziv1234 committed:
first working version - need tests

ziv1234 committed:
added unit-tests

ziv1234 commented:
adding a comment so it does not auto close. nothing to add until it is reviewed :)

ziv1234 commented:
had a merge conflict in manifest.json from another PR when it was approved. now all good

ziv1234 committed:
fixed isort from merge

ziv1234 committed:
fixed hassfest merge conflict error

ziv1234 committed:
fixed client 'id' to int type. Aparently there is a new check

ziv1234 commented:
hello. I rebased it again and fixed the conflicts. can someone please review this? it has been sitting for a very long time with no comments

farmio reviewed (changes requested):
I did only see some minor points, but I'm not very familiar with websockets and panels yet 😬

farmio commented:
Oh, Is the yaml config not deprecated here? I guess it would make sense since when a user changes / reconfigures a bridge via the new panel, that is also configured via yaml it would not persist those changes, would it?

farmio commented on homeassistant/components/dynalite/panel.py:
```diff
@@ -0,0 +1,117 @@
+        path = locate_dir()
+        build_id = get_build_id()
+        hass.http.register_static_path(
+            URL_BASE, path, cache_headers=(build_id == "dev")
```
Why would you cache in "dev", but not in production?

ziv1234 replied:
you are correct. changed to using default value (true)

farmio replied:
I'd use `build_id != "dev"`

ziv1234 replied:
done

ziv1234 commented:
apologies. i have been out for a few weeks. will get on it quickly

ziv1234 committed:
simplification based on farmio suggestion

ziv1234 committed:
fix - cache headers when not dev (farmio)

farmio commented:
Still not sure about this: https://github.com/home-assistant/core/pull/77181#issuecomment-1460996027
Is YAML still supported? Is it imported? 

If imported and converted, I think it should be deprecated too, and a repair issue should be created for the user to remove stale yaml config then.

ziv1234 commented:
> Still not sure about this: [#77181 (comment)](https://github.com/home-assistant/core/pull/77181#issuecomment-1460996027) Is YAML still supported? Is it imported?
> 
> If imported and converted, I think it should be deprecated too, and a repair issue should be created for the user to remove stale yaml config then.

the yaml config is still imported and converted. I did this, so it would not be a breaking change. do you have an example of a depracation / repair issue? i will check in knx as well

farmio commented:
Knx doesn't import yaml. 
Have a look at any of the recent release notes. You'll find a section "configurable from UI". There you can find some integrations that did such conversion and also a standard text for that in their PR description.

ziv1234 committed:
add an issue when deprecated YAML configuration
still exists

ziv1234 commented:
> Knx doesn't import yaml. Have a look at any of the recent release notes. You'll find a section "configurable from UI". There you can find some integrations that did such conversion and also a standard text for that in their PR description.

thanks. implemented per your suggestion from bratt...

farmio reviewed (commented):
Just a minor addition - not sure if this is required, but can't hurt.

farmio reviewed (commented):
_post was duplicated_

ziv1234 committed:
added deprecation warning

ziv1234 commented on homeassistant/components/dynalite/__init__.py:
```diff
@@ -17,158 +14,19 @@
+from .panel import async_register_dynalite_frontend
+from .schema import BRIDGE_SCHEMA
 
 CONFIG_SCHEMA = vol.Schema(
```
thanks. added and pushed

ziv1234 commented:
implemented the last change

farmio reviewed (approved):
This looks good to me! 👍
I have tested it locally to see the panel in action (mocking the config entry since not having Dynalite equipment) and it seems to work as intended.

ziv1234 committed:
changed version number of the panel
no real changes, but version number was messed up

elupus commented on homeassistant/components/dynalite/panel.py:
```diff
@@ -0,0 +1,117 @@
+        return
+    message_conf = msg["config"]
+    message_data = {
+        conf: message_conf[conf] for conf in RELEVANT_CONFS if conf in message_conf
```
Should it perhaps validate config before saving?

ziv1234 replied:
not sure I understand. the schema is being validated by the websocket api (homeassistant/components/websocket_api/connection.py line 200) just before the callback. I can run it through the schema again but I believe it will only add complexity. in addition, how would i test it? homeassistant will not forward a command that fails the schema.

elupus replied:
Ah okey. So the config will be validate by websocket. Then its fine.
